copulas/multivariate/Tree.py

```diff
@@ -0,0 +1,379 @@
+            U2givenU1 = derivative(U1, U2, copula_para)
+            # correction of 0 or 1'''
+            U1givenU2[U1givenU2 == 0], U2givenU1[U2givenU1 == 0] = eps, eps
+            U1givenU2[U1givenU2 == 1], U2givenU1[U2givenU1 == 1] = 1 - eps, 1 - eps
+            edge.U = U1givenU2
+
+    def _likehood_T(self, U):
+        """Compute likelihood of the tree given an U matrix
+        """
+        newU = np.empty([self.vine.n_var, self.vine.n_var])
+        edge_set = self.edge_set
+        values = np.zeros([1, len(edge_set)])
+        for i in range(len(edge_set)):
+            edge = edge_set[i]
+            cname = self.vine.c_map[edge.name]
+            v1 = edge.L
+            v2 = edge.R
+            copula_para = edge.param
+            if self.level == 1:
+                U1 = np.array([U[v1]])
+                U2 = np.array([U[v2]])
+            else:
+                prev_T = self.prev_T.edge_set
+                l_p = edge.parent[0]
+                r_p = edge.parent[1]
+                U1, U2 = prev_T[l_p].U, prev_T[r_p].U
+            cop = Copula(cname=cname)
+            cop.fit(U1, U2)
+            pdf = cop.get_pdf()
+            derivative = cop.get_h_function()
+            values[0, i] = pdf(U1, U2)
+            U1givenU2 = derivative(U2, U1, copula_para)
+            # U2givenU1 = derivative(U1, U2, copula_para)
+            edge.U = U1givenU2
+            value = np.sum(np.log(values))
+        return newU, value
+
+    def print_tree(self):
+        for e in self.edge_set:
+            msg = "L:{} R:{} D:{} parent:{}".format(e.L, e.R, e.D, e.parent)
+            LOGGER.info(msg)
+
+
+class CTree(Tree):
+    def __init__(self, k, n, tau_mat, prev_T):
+        super(CTree, self).__init__(k, n, tau_mat, prev_T)
+
+    def _build_first_tree(self):
+        # first column is the variable of interest
+        tau_mat = self.tau_mat
+        np.fill_diagonal(tau_mat, np.NaN)
+        tau_y = tau_mat[:, 0]
+        temp = np.empty([self.n_nodes, 3])
+        temp[:, 0] = np.arange(self.n_nodes)
+        temp[:, 1] = tau_y
+        temp[:, 2] = abs(tau_y)
+        temp[np.isnan(temp)] = -10
+        tau_sorted = temp[temp[:, 2].argsort()[::-1]]
+        for itr in range(1, self.n_nodes):
+            ind = tau_sorted[itr, 0]
+            name, param = Copula.select_copula(self.u_matrix[:, 0],
+                                               self.u_matrix[:, ind])
+            new_edge = Edge(itr, 0, ind, tau_mat[0, ind], name, param)
+            new_edge.parent.append(0)
+            new_edge.parent.append(ind)
+            self.edge_set.append(new_edge)
+
+    def _build_kth_tree(self):
+        # find anchor variable with highest sum of dependence with the rest
+        temp = np.empty([self.n_nodes, 2])
+        temp[:, 0] = np.arange(self.n_nodes, dtype=int)
+        temp[:, 1] = np.sum(abs(self.tau_mat), 1)
+        tau_sorted = temp[temp[:, 1].argsort()[::-1]]
+        anchor = int(temp[0, 0])
```

Comment:
The lines 185:189 should be refactored out in a function `get_anchor`. This way, the function could start with:
```python
anchor = self.get_anchor()
```